### PR TITLE
Refactored allocator and resolved a few scheduling race conditions

### PR DESCRIPTION
**Pull request by ellistarn**
```
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:28:02.403Z	INFO	Found 1 provisionable pods
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:28:08.712Z	DEBUG	Successfully discovered launch template Karpenter-etarn-dev for cluster etarn-dev
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:28:08.712Z	DEBUG	Successfully packed 1 pods onto 1 nodes
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:28:08.891Z	DEBUG	Successfully discovered subnets in 3 zones for cluster etarn-dev
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:28:10.870Z	INFO	Binding 1 pods to node ip-192-168-142-252.us-west-2.compute.internal
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:28:10.887Z	DEBUG	Successfully bound pod default/pause-5d968bbb58-wfrg8 to node ip-192-168-142-252.us-west-2.compute.internal
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:28:55.970Z	INFO	Found 2 provisionable pods
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:28:55.970Z	DEBUG	Successfully packed 1 pods onto 1 nodes
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:28:57.788Z	DEBUG	Successfully packed 1 pods onto 1 nodes
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:28:59.529Z	DEBUG	Retrying DescribeInstances due to eventual consistency: fleet created, but instances not yet found.
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:28:59.956Z	INFO	Binding 1 pods to node ip-192-168-89-35.us-west-2.compute.internal
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:28:59.971Z	DEBUG	Successfully bound pod default/pause-5d968bbb58-gv6kb to node ip-192-168-89-35.us-west-2.compute.internal
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:28:59.971Z	INFO	Binding 1 pods to node ip-192-168-130-200.us-west-2.compute.internal
karpenter-9cc5bc955-8f8dk manager 2021-02-04T21:29:00.014Z	ERROR	Continuing after failing to bind, binding pod, Operation cannot be fulfilled on pods/binding "pause-5d968bbb58-9s6zq": pod pause-5d968bbb58-9s6zq is already assigned to node "ip-192-168-89-35.us-west-2.compute.internal"
```

### CODE REVIEW

**Review by prateekgogia — APPROVED** [review]

**Comment by ellistarn** [comment]
Gonna hold off on this merge until the other is merged

File: pkg/controllers/provisioning/v1alpha1/allocator/filter.go
Comment (prateekgogia) [inline pkg/controllers/provisioning/v1alpha1/allocator/filter.go:38]: Shall we call it `GetPendingPods` or `GetUnschedulablePods`

**Review by prateekgogia — APPROVED** [review]

File: pkg/controllers/provisioning/v1alpha1/allocator/bind.go
Comment (njtran) [inline pkg/controllers/provisioning/v1alpha1/allocator/bind.go:58]: What do you think about using
`KarpenterNodeLabelKey: fmt.Sprintf("Karpenter.%s.%s", provisioner.Name, provisioner.Namespace),`

since other labels use periods in between and are prefixed for where they come from?

File: pkg/cloudprovider/types.go
Comment (prateekgogia) [inline pkg/cloudprovider/types.go:85]: nit: packing

File: pkg/controllers/provisioning/v1alpha1/allocator/constraints.go
Comment (njtran) [inline pkg/controllers/provisioning/v1alpha1/allocator/constraints.go:63]: Do we want to default to Linux here? or is this just a placeholder?

**Review by prateekgogia — APPROVED** [review]

File: pkg/controllers/provisioning/v1alpha1/allocator/constraints.go
Comment (njtran) [inline pkg/controllers/provisioning/v1alpha1/allocator/constraints.go:38]: Do you need to include this before the groups for loop? If not, the first pod iteration will always be left out.

**Review by prateekgogia — APPROVED** [review]

**Review by njtran — COMMENTED** [review]
Looks good!

File: pkg/controllers/provisioning/v1alpha1/allocator/constraints.go
Comment (ellistarn) [reply pkg/controllers/provisioning/v1alpha1/allocator/constraints.go:63]: This is just refactored code. We are defaulting to linux, for now.

File: pkg/controllers/provisioning/v1alpha1/allocator/constraints.go
Comment (ellistarn) [reply pkg/controllers/provisioning/v1alpha1/allocator/constraints.go:38]: I'm gonna take another pass at this. This is copied from greedyallocator.

**Pull request merged**
